src/shogun/evaluation/CrossValidatedCalibration.cpp

```diff
@@ -0,0 +1,399 @@
+#include <shogun/labels/LabelsFactory.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/mathematics/Statistics.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+
+using namespace shogun;
+using namespace linalg;
+
+CCrossValidatedCalibration::CCrossValidatedCalibration() : CMachine()
+{
+	init();
+}
+
+CCrossValidatedCalibration::CCrossValidatedCalibration(
+    CMachine* machine, CLabels* labels, CSplittingStrategy* splitting_strategy,
+    CCalibrationMethod* calibration_method)
+    : CMachine()
+{
+	init();
+
+	m_machine = machine;
+	m_labels = labels;
+	m_splitting_strategy = splitting_strategy;
+	m_calibration_method = calibration_method;
+
+	SG_REF(m_machine)
+	SG_REF(m_labels)
+	SG_REF(m_splitting_strategy)
+	SG_REF(m_calibration_method)
+}
+
+CCrossValidatedCalibration::~CCrossValidatedCalibration()
+{
+	SG_UNREF(m_machine);
+	SG_UNREF(m_labels);
+	SG_UNREF(m_splitting_strategy);
+	SG_UNREF(m_calibration_method);
+	SG_UNREF(m_calibration_machines);
+}
+
+void CCrossValidatedCalibration::init()
+{
+	m_machine = new CMachine();
+	m_labels = new CBinaryLabels();
+	m_splitting_strategy = new CStratifiedCrossValidationSplitting();
+	m_calibration_method = new CCalibrationMethod();
+	m_calibration_machines = new CDynamicObjectArray();
+
+	SG_ADD(
+	    (CSGObject**)&m_machine, "m_machine", "machine to be calibrated",
+	    MS_NOT_AVAILABLE);
+	SG_ADD(
+	    (CSGObject**)&m_labels, "m_labels", "target labels", MS_NOT_AVAILABLE);
+	SG_ADD(
+	    (CSGObject**)&m_splitting_strategy, "m_splitting_strategy",
+	    "splitting strategy for cross-validation", MS_NOT_AVAILABLE);
+	SG_ADD(
+	    (CSGObject**)&m_calibration_method, "m_calibration_method",
+	    "method for calibrating the predictions", MS_NOT_AVAILABLE);
+	SG_ADD(
+	    (CSGObject**)&m_calibration_machines, "m_calibration_machines",
+	    "array of calibration machines", MS_NOT_AVAILABLE);
+}
+
+EProblemType CCrossValidatedCalibration::get_machine_problem_type() const
+{
+	return m_machine->get_machine_problem_type();
+}
+
+bool CCrossValidatedCalibration::train(CFeatures* data)
+{
+	// code borrowed from Calibration.cpp
+	index_t num_subsets = m_splitting_strategy->get_num_subsets();
+
+	m_calibration_machines = new CDynamicObjectArray(num_subsets);
+
+	SG_DEBUG(
+	    "building index sets for %d-fold cross-validated calibration\n",
+	    num_subsets)
+
+	/* build index sets */
+	m_splitting_strategy->build_subsets();
+
+	if (m_machine->is_data_locked())
+	{
+		SG_ERROR("cannot run train on locked data use train_locked instead\n")
+		return false;
+	}
+	SG_DEBUG("starting unlocked calibration\n", get_name())
+	/* tell machine to store model internally
+	 * (otherwise changing subset of features will kaboom the classifier) */
+	m_machine->set_store_model_features(true);
+
+	for (index_t i = 0; i < num_subsets; ++i)
+	{
+		CMachine* machine;
+		CFeatures* features;
+		CLabels* labels;
+
+		if (get_global_parallel()->get_num_threads() == 1)
```

Comment:
Just realised we really don't want to copy/replicate this buggy x-validation codes ...